pep-0xxx.rst

```diff
@@ -0,0 +1,571 @@
+``TLSWrappedBuffer`` objects.
+
+The ``Context`` abstract base class has the following class definition::
+
+    TLSBufferObject = Union[TLSWrappedSocket, TLSWrappedBuffer]
+    ServerNameCallback = Callable[[TLSBufferObject, Optional[str], Context], Any]
+
+    class Context(metaclass=ABCMeta):
+        @abstractmethod
+        def register_certificates(self,
+                                  certificates: str,
+                                  key=None: Optional[str],
+                                  password=None: Optional[Callable[[], Union[AnyStr, bytearray]]]) -> None:
+            """
+            Loads a certificate, a number of intermediate certificates, and the
+            corresponding private key. These certificates will be offered to
+            the remote peer during the handshake if required.
+
+            The ``certificates`` argument must be a bytestring containing the
+            PEM-encoded certificates. The first PEM-encoded certificate must be
+            the leaf certificate. All subsequence certificates will be offered
+            as intermediate additional certificates.
+
+            The ``key`` argument, if present, must contain the PEM-encoded
+            private key associated with the leaf certificate. If not present,
+            the private key will be extracted from ``certificates``.
+
+            The ``password`` argument may be a function to call to get the
+            password for decrypting the private key. It will only be called if
+            the private key is encrypted and a password is necessary. It will
+            be called with no arguments, and it should return a string, bytes,
+            or bytearray. If the return value is a string it will be encoded as
+            UTF-8 before using it to decrypt the key. Alternatively a string,
+            bytes, or bytearray value may be supplied directly as the password
+            argument. It will be ignored if the private key is not encrypted
+            and no password is needed.
+            """
+            pass
+
+        @abstractmethod
+        def set_ciphers(self, ciphers: List[Ciphers]) -> None:
+            """
+            Set the available ciphers for TLS connections created with this
+            context. ``ciphers`` should be a list of ciphers from the
+            ``Cipher`` registry. If none of the ``ciphers`` provided to this
+            object are supported or available, a ``TLSError`` will be raised.
+            """
+            pass
+
+        @abstractmethod
+        def set_inner_protocols(self, protocols: List[str]) -> None:
+            """
+            Specify which protocols the socket should advertise as supported
+            during the TLS handshake. This may be advertised using either or
+            both of ALPN or NPN.
+
+            ``protocols`` should be a list of ASCII strings for ALPN tokens,
+            such as ``['h2', 'http/1.1']``, ordered by preference. The
```

Comment:
How about having one generic type that eats a string?  Like:

```python
protocols = [H2, HTTP_1_1, Custom("whatever")]
```

Having proper objects for well-known protocols seems like a good way to avoid (typo) errors and avoiding normalization (like `'h2 '` or `'H2'`).